File: node/src/models/DataInterpreter.ts
import Group from './Group'
import { cartesian } from '../utils';

export default class DataInterpreter {
  private _rows: Group[];
  private _columns: Group[];
  private _squares: Group[];

  constructor(private _data: number[][]) {
    this._rows = this._data.map(row => new Group(row))

    this._columns = [];
    for (let x = 0; x < 9; x++) {
      let data = []
      for (let y = 0; y < 9; y++) {
        data.push(this._data[y][x])
      }
      this._columns.push(new Group(data))
    }

    this._squares= [];
    [[0,1,2], [3,4,5], [6,7,8]].forEach((a) => {
      [[0,1,2], [3,4,5], [6,7,8]].forEach((b) => {
        let data = cartesian(a, b).map((xy: number[]) => this._data[xy[0]][xy[1]])
        this._squares.push(new Group(data))
      })
    })
  }

  get rows(): Group[] {
    return this._rows
  }

  get columns(): Group[] {
    return this._columns
  }

  get squares(): Group[] {
    return this._squares
  }
}

File: c++/data_interpreter.cpp
#include <vector>
#include <string>
#include "data_interpreter.h"
#include "group.h"

using namespace std;

DataInterpreter::DataInterpreter(const vector<vector<int>>) {
  // rows = data.map { |row| Group.new(row) }
  rows = []
  for (int i = 0; i < 9, i++) {

  }




  // columns = (0...9).map { |x| Group.new((0...9).map { |y| @data[y][x] }) }

  // squares = []
  // [[0,1,2], [3,4,5], [6,7,8]].each do |a|
  //   [[0,1,2], [3,4,5], [6,7,8]].each do |b|
  //     squares << Group.new(a.product(b).map { |x, y| @data[x][y] })
  //   end
  // end
}
